packages/fuels-contract/src/contract.rs

```diff
@@ -563,30 +384,22 @@ where
     /// your method returns `bool`, it will be a bool, works also for structs thanks to the
     /// `abigen!()`). The other field of CallResponse, `receipts`, contains the receipts of the
     /// transaction.
+    #[tracing::instrument]
     async fn call_or_simulate(self, simulate: bool) -> Result<CallResponse<D>, Error> {
-        let receipts = Contract::call(
-            self.contract_id,
-            Some(self.encoded_selector),
-            Some(self.encoded_args),
-            &self.fuel_client,
-            self.tx_parameters,
-            self.call_parameters,
-            self.variable_outputs,
-            self.maturity,
-            self.compute_calldata_offset,
-            self.external_contracts,
-            self.wallet,
-            simulate,
-        )
-        .await?;
+        let script = self.build_script().await;
 
-        // If it's an ABI method without a return value, exit early.
-        if self.output_params.is_empty() {
-            return Ok(CallResponse::new(D::from_tokens(vec![])?, receipts));
-        }
+        let receipts = if simulate {
+            script.simulate(&self.fuel_client).await?
+        } else {
+            script.call(&self.fuel_client).await?
+        };
+        tracing::debug!(target: "receipts", "{:?}", receipts);
 
-        let (decoded_value, receipts) = Self::get_decoded_output(receipts, &self.output_params)?;
-        Ok(CallResponse::new(D::from_tokens(decoded_value)?, receipts))
+        self.build_response(receipts)
```

Comment:
If the `build_response` function returns a response, the method should be called `get_response` or something like that. Otherwise it looks like a builder method.